app/src/main/java/org/schabi/newpipe/local/playlist/LocalPlaylistFragment.java

```diff
@@ -331,6 +358,121 @@ public void onComplete() { }
         };
     }
 
+    @Override
+    public boolean onOptionsItemSelected(final MenuItem item) {
+        switch (item.getItemId()) {
+            case R.id.menu_item_removeWatched:
+                android.app.AlertDialog.Builder builder =
+                new android.app.AlertDialog.Builder(getActivity());
+                builder.setMessage(R.string.remove_watched_popup_warning)
+                        .setTitle(R.string.remove_watched_popup_title)
+                        .setPositiveButton(R.string.remove_watched_popup_yes,
+                                (DialogInterface d, int id) -> removeWatchedStreams(false))
+                        .setNeutralButton(
+                                R.string.remove_watched_popup_yes_and_partially_watched_videos,
+                                (DialogInterface d, int id) -> removeWatchedStreams(true))
+                        .setNegativeButton(R.string.remove_watched_popup_cancel,
+                                (DialogInterface d, int id) -> d.cancel());
+                builder.create().show();
+                break;
+            default:
+                return super.onOptionsItemSelected(item);
+        }
+        return true;
+    }
+
+    public void removeWatchedStreams(final boolean removePartiallyWatched) {
+        if (removeWatchedDisposable != null && !removeWatchedDisposable.isDisposed()) {
+            // already running
+            return;
+        }
+        showLoading();
+
+        removeWatchedDisposable = Flowable.just(Flowable.just(removePartiallyWatched,
+                playlistManager.getPlaylistStreams(playlistId).blockingFirst()))
+                .subscribeOn(Schedulers.io())
+                .map(flow -> {
+                    boolean localRemovePartiallyWatched = (boolean) flow.blockingFirst();
+                    List<PlaylistStreamEntry> playlist
+                            = (List<PlaylistStreamEntry>) flow.blockingLast();
+                    HistoryRecordManager recordManager = new HistoryRecordManager(getContext());
+                    Iterator<PlaylistStreamEntry> playlistIter = playlist.iterator();
+                    Iterator<StreamHistoryEntry> historyIter = recordManager
+                            .getStreamHistorySortedById().blockingFirst().iterator();
+                    List<PlaylistStreamEntry> notWatchedItems = new ArrayList<>();
+                    Iterator<StreamStateEntity> streamStatesIter = null;
+                    boolean thumbnailVideoRemoved = false;
```

Comment:
In my opinion it is difficult to read all of these declarations here, maybe they should be distributed right befor the places in code where they are being used (e.g. `streamStatesIter` would be moved on line 405, before the `if (!localRemovePartiallyWatched)` branch where it is used). But I'm not sure if it's the same for others. ;-)

Reply:
How does this look?
![code](https://user-images.githubusercontent.com/50427197/79401234-2ef1da00-7f4e-11ea-9709-a07942fe78b2.png)

Reply:
Looks better, thanks :-D

Reply:
Fixed in db335d5